### DISCUSSION TITLE
Uneven VDEVs and data distribution

### DISCUSSION THREAD

**Discussion by filonovd** [openzfs/zfs#13449] [General]
Hi - 

 As far as I understand ZFS should distribute data according to the percentage of available space on each VDEV. Sounds really great. On paper. 
Yesterday I had a chance to test it out in production by building from scratch a fairly large backup zpool. I had 44x14TB+18x10TB+16x12TB drives, so I came up with this config - 

```
# zpool list -v
NAME        SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
zem         899T  48.7T   850T        -         -     0%     5%  1.00x    ONLINE  -
  raidz2    140T  3.56T   137T        -         -     0%  2.54%      -  ONLINE  
    da0        -      -      -        -         -      -      -      -  ONLINE  
    da1        -      -      -        -         -      -      -      -  ONLINE  
    da2        -      -      -        -         -      -      -      -  ONLINE  
    da3        -      -      -        -         -      -      -      -  ONLINE  
    da4        -      -      -        -         -      -      -      -  ONLINE  
    da5        -      -      -        -         -      -      -      -  ONLINE  
    da6        -      -      -        -         -      -      -      -  ONLINE  
    da7        -      -      -        -         -      -      -      -  ONLINE  
    da8        -      -      -        -         -      -      -      -  ONLINE  
    da9        -      -      -        -         -      -      -      -  ONLINE  
    da10       -      -      -        -         -      -      -      -  ONLINE  
  raidz2    140T  3.43T   137T        -         -     0%  2.44%      -  ONLINE  
    da11       -      -      -        -         -      -      -      -  ONLINE  
    da12       -      -      -        -         -      -      -      -  ONLINE  
    da13       -      -      -        -         -      -      -      -  ONLINE  
    da14       -      -      -        -         -      -      -      -  ONLINE  
    da15       -      -      -        -         -      -      -      -  ONLINE  
    da16       -      -      -        -         -      -      -      -  ONLINE  
    da17       -      -      -        -         -      -      -      -  ONLINE  
    da18       -      -      -        -         -      -      -      -  ONLINE  
    da19       -      -      -        -         -      -      -      -  ONLINE  
    da20       -      -      -        -         -      -      -      -  ONLINE  
    da21       -      -      -        -         -      -      -      -  ONLINE  
  raidz2    140T  3.17T   137T        -         -     0%  2.26%      -  ONLINE  
    da22       -      -      -        -         -      -      -      -  ONLINE  
    da23       -      -      -        -         -      -      -      -  ONLINE  
    da24       -      -      -        -         -      -      -      -  ONLINE  
    da25       -      -      -        -         -      -      -      -  ONLINE  
    da26       -      -      -        -         -      -      -      -  ONLINE  
    da27       -      -      -        -         -      -      -      -  ONLINE  
    da28       -      -      -        -         -      -      -      -  ONLINE  
    da29       -      -      -        -         -      -      -      -  ONLINE  
    da30       -      -      -        -         -      -      -      -  ONLINE  
    da31       -      -      -        -         -      -      -      -  ONLINE  
    da32       -      -      -        -         -      -      -      -  ONLINE  
  raidz2    140T  3.12T   137T        -         -     0%  2.22%      -  ONLINE  
    da33       -      -      -        -         -      -      -      -  ONLINE  
    da34       -      -      -        -         -      -      -      -  ONLINE  
    da35       -      -      -        -         -      -      -      -  ONLINE  
    da36       -      -      -        -         -      -      -      -  ONLINE  
    da37       -      -      -        -         -      -      -      -  ONLINE  
    da38       -      -      -        -         -      -      -      -  ONLINE  
    da39       -      -      -        -         -      -      -      -  ONLINE  
    da40       -      -      -        -         -      -      -      -  ONLINE  
    da41       -      -      -        -         -      -      -      -  ONLINE  
    da42       -      -      -        -         -      -      -      -  ONLINE  
    da43       -      -      -        -         -      -      -      -  ONLINE  
  raidz2   81.9T  8.53T  73.3T        -         -     0%  10.4%      -  ONLINE  
    da46       -      -      -        -         -      -      -      -  ONLINE  
    da47       -      -      -        -         -      -      -      -  ONLINE  
    da48       -      -      -        -         -      -      -      -  ONLINE  
    da49       -      -      -        -         -      -      -      -  ONLINE  
    da50       -      -      -        -         -      -      -      -  ONLINE  
    da51       -      -      -        -         -      -      -      -  ONLINE  
    da52       -      -      -        -         -      -      -      -  ONLINE  
    da53       -      -      -        -         -      -      -      -  ONLINE  
    da54       -      -      -        -         -      -      -      -  ONLINE  
  raidz2   81.9T  8.52T  73.3T        -         -     0%  10.4%      -  ONLINE  
    da55       -      -      -        -         -      -      -      -  ONLINE  
    da56       -      -      -        -         -      -      -      -  ONLINE  
    da57       -      -      -        -         -      -      -      -  ONLINE  
    da58       -      -      -        -         -      -      -      -  ONLINE  
    da59       -      -      -        -         -      -      -      -  ONLINE  
    da60       -      -      -        -         -      -      -      -  ONLINE  
    da61       -      -      -        -         -      -      -      -  ONLINE  
    da62       -      -      -        -         -      -      -      -  ONLINE  
    da63       -      -      -        -         -      -      -      -  ONLINE  
  raidz2   87.3T  8.37T  78.9T        -         -     0%  9.58%      -  ONLINE  
    da64       -      -      -        -         -      -      -      -  ONLINE  
    da65       -      -      -        -         -      -      -      -  ONLINE  
    da66       -      -      -        -         -      -      -      -  ONLINE  
    da67       -      -      -        -         -      -      -      -  ONLINE  
    da68       -      -      -        -         -      -      -      -  ONLINE  
    da69       -      -      -        -         -      -      -      -  ONLINE  
    da70       -      -      -        -         -      -      -      -  ONLINE  
    da71       -      -      -        -         -      -      -      -  ONLINE  
  raidz2   87.3T  9.99T  77.3T        -         -     0%  11.4%      -  ONLINE  
    da72       -      -      -        -         -      -      -      -  ONLINE  
    da73       -      -      -        -         -      -      -      -  ONLINE  
    da74       -      -      -        -         -      -      -      -  ONLINE  
    da75       -      -      -        -         -      -      -      -  ONLINE  
    da76       -      -      -        -         -      -      -      -  ONLINE  
    da77       -      -      -        -         -      -      -      -  ONLINE  
    da78       -      -      -        -         -      -      -      -  ONLINE  
    da79       -      -      -        -         -      -      -      -  ONLINE  
```

As you can see the largest VDEVs (four 11x14TB raidz2) are the least used. And the smaller ones are being used more. 
Can somebody explain what's going on here? Am still pushing data there and will see how that data distributes across the pool later on. But if there's something that I didn't do right to have it distributed properly then it's better to learn now and have it fixed before I have whole 500TB backed up. 

P.S. The data files are fairly small, it's not like I pushed a single file that is 8TB in size and it just landed on an empty VDEV.

**Comment by amotin** [comment]
ZFS indeed tries to balance the write proportional to the use percentage of specific vdev vs the whole pool.  But this logic fights with other logic, that tries to send more writes to vdev that can write faster.  So for slow writes the first logic shod prevail, but for fast -- likely the second. Is there chance that your smaller drives are somehow much faster?

PS: If you are using large blocks, then the first logic may also be less efficient, and this change may slightly improve situation: https://github.com/openzfs/zfs/pull/13388 .

**Reply by filonovd** [reply]
Thanks for getting back to me. 
The newer and bigger drives are connected via SAS expander which might slow them down overall. Is there any tunables to give less priority to the "speed" logic versus "space" logic? 
Using faster VDEVs is fine for as long as we have space there. But soon these will be full and then it will have to use slower ones and only slower ones. Which sounds a bit sub-optimal. On the other hand - this is pure backup server and access speed is not that important there.

**Comment by minorsatellite** [comment]
Sampling the three different systems under my direction, I can say that
VDEV allocation can vary pretty wildly. An early FreeBSD and more recently
patched ZOL system running on Ubuntu 20.04LTS look fairly balanced, but the
other Ubuntu ZOL 20.04 LTS system not so much. All disk and VDEV groups are
the same, all fabrics SAS3.

Below is the outlier Linux system

zpool list -v
> NAME                         SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG
>    CAP  DEDUP    HEALTH  ALTROOT
> pool3                        254T   163T  91.6T        -         -     1%
>    64%  1.00x    ONLINE  -
>   raidz3                     127T  90.6T  36.4T        -         -     3%
>  71.4%      -  ONLINE
>     wwn-0x5000cca264340cec      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26434359c      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26433cf48      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2646d917c      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26430cfbc      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26400738c      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26400f3e0      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26400dc90      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca26400dd90      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2646b9fbc      -      -      -        -         -      -
>      -      -  ONLINE
>   raidz3                     127T  72.1T  55.2T        -         -     0%
>  56.7%      -  ONLINE
>     wwn-0x5000cca28e899370      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca28e8bc248      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca28e8bf3d0      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2973ac288      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2973b953c      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2973ba8a4      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2973c7360      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca29816b850      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2a37672f8      -      -      -        -         -      -
>      -      -  ONLINE
>     wwn-0x5000cca2a387d940      -      -      -        -         -      -
>      -      -  ONLINE
